test/opt/test_unify_const.cpp

```diff
@@ -0,0 +1,754 @@
+          "%unsigned_1 = OpConstant %uint 1",
+      })
+      .AppendInMain({
+          "%uint_var = OpVariable %_pf_uint Function",
+          "OpStore %uint_var %unsigned_1",
+      })
+      .AppendNames({
+          "OpName %unsigned_1 \"unsigned_1_duplicate\"",  // the OpName
+                                                          // instruction of the
+                                                          // removed duplicated
+                                                          // constant won't be
+                                                          // erased.
+      });
+  Check();
+}
+
+TEST_F(UnifyFrontEndConstantSingleTest, SkipWhenResultIdHasDecorations) {
+  test_builder_
+      .AppendAnnotations({
+          // So far we don't have valid decorations for constants. This is
+          // preparing for the future updates of SPIR-V.
+          // TODO(qining): change to a valid decoration once they are available.
+          "OpDecorate %s_duplicate RelaxedPrecision",
+      })
+      .AppendTypesConstantsGlobals({
+          // clang-format off
+          "%bool_true = OpConstantTrue %bool",
+          "%signed_1 = OpConstant %int 1",
+          "%float_1 = OpConstant %float 1",
+          "%double_1 = OpConstant %double 1",
+          "%s = OpConstantComposite %flat_struct %bool_true %signed_1 %float_1 %double_1",
+          // The duplicated constants without decorations should not be skipped.
+          "%float_1_duplicate = OpConstant %float 1",
+          "%double_1_duplicate = OpConstant %double 1",
+          // The type "flat_struct" has decoration, so the following
+          // instruction should not be removed.
+          "%s_duplicate = OpConstantComposite %flat_struct %bool_true %signed_1 %float_1_duplicate %double_1_duplicate",
+          // clang-format on
+      })
+      .AppendInMain({
+          "%s_var = OpVariable %_pf_flat_struct Function",
+          "OpStore %s_var %s_duplicate",
+      });
+
+  expected_builder_
+      .AppendAnnotations({
+          "OpDecorate %s_duplicate RelaxedPrecision",
+      })
+      .AppendTypesConstantsGlobals({
+          // clang-format off
+          "%bool_true = OpConstantTrue %bool",
+          "%signed_1 = OpConstant %int 1",
+          "%float_1 = OpConstant %float 1",
+          "%double_1 = OpConstant %double 1",
+          "%s = OpConstantComposite %flat_struct %bool_true %signed_1 %float_1 %double_1",
+          "%s_duplicate = OpConstantComposite %flat_struct %bool_true %signed_1 %float_1 %double_1",
+          // clang-format on
+      })
+      .AppendInMain({
+          "%s_var = OpVariable %_pf_flat_struct Function",
+          "OpStore %s_var %s_duplicate",
+      })
+      .AppendNames({
+          "OpName %float_1 \"float_1_duplicate\"",
+          "OpName %double_1 \"double_1_duplicate\"",
+      });
+  Check();
+}
+
+struct UnifyConstantTestCase {
+  // new constants that should be kept.
+  std::vector<std::string> new_consts;
```

Comment:
Instead of `new_consts` and `use_new_consts`, what about `preserved_consts` and `use_preserved_consts`? I feel the later is more clear.